concordium-cis2/src/lib.rs

```diff
@@ -1120,6 +1132,156 @@ pub struct OnReceivingCis2Params<T: IsTokenId, A: IsTokenAmount> {
+#[derive(Debug, Serial, PartialEq, Eq)]
+pub struct StandardIdentifier<'a> {
+    #[concordium(size_length = 1)]
+    id: &'a str,
+}
+
+/// String is invalid as standard identifier. Ensure the length is less than 256
+/// and it only contains ASCII characters.
+#[derive(Default)]
+pub struct InvalidStandardIdentifierError;
+
+impl<'a> StandardIdentifier<'a> {
+    /// Validate and construct a standard identifier.
+    pub fn new(id: &'a str) -> Result<Self, InvalidStandardIdentifierError> {
+        if id.len() > 255 || !id.is_ascii() {
+            Err(InvalidStandardIdentifierError)
+        } else {
+            Ok(Self {
+                id,
+            })
+        }
+    }
+
+    /// Construct a standard identifier without validation.
+    pub const fn new_unchecked(id: &'a str) -> Self {
+        Self {
+            id,
+        }
+    }
+
+    /// Convert to owned standard identifier.
+    pub fn to_owned(&self) -> StandardIdentifierOwned {
+        StandardIdentifierOwned::new_unchecked(self.id.to_string())
+    }
+}
+
+/// Owned identifier for a smart contract standard.
+/// Consists of a string of ASCII characters up to a length of 255.
+///
+/// See [StandardIdentifier] for the borrowed version.
+#[derive(Debug, Serialize, PartialEq, Eq)]
+pub struct StandardIdentifierOwned {
+    #[concordium(size_length = 1)]
+    id: String,
+}
+
+impl schema::SchemaType for StandardIdentifierOwned {
+    fn get_type() -> schema::Type { schema::Type::String(schema::SizeLength::U8) }
+}
+
+impl StandardIdentifierOwned {
+    /// Validate and construct a standard identifier.
+    pub fn new(id: String) -> Result<Self, InvalidStandardIdentifierError> {
+        if id.len() > 255 || !id.is_ascii() {
+            Err(InvalidStandardIdentifierError)
+        } else {
+            Ok(Self {
+                id,
+            })
+        }
+    }
+
+    /// Construct a standard identifier without validation.
+    pub fn new_unchecked(id: String) -> Self {
+        Self {
+            id,
+        }
+    }
+
+    /// Convert to standard identifier.
+    pub fn as_standard_identifier(&self) -> StandardIdentifier {
+        StandardIdentifier::new_unchecked(&self.id)
+    }
+}
+
+/// The parameter type for the contract function `supports`.
+#[derive(Debug, Serialize)]
+pub struct SupportsQueryParams {
+    /// The list of support queries.
+    #[concordium(size_length = 2)]
+    pub queries: Vec<StandardIdentifierOwned>,
+}
+
+impl schema::SchemaType for SupportsQueryParams {
+    fn get_type() -> schema::Type {
+        schema::Type::List(schema::SizeLength::U16, Box::new(StandardIdentifierOwned::get_type()))
+    }
+}
+
+/// The query result type for whether a smart contract supports a standard.
+// Note: For the serialization to be derived according to the CIS0
+// specification, the order of the variants and their fields cannot be changed.
+#[derive(Debug, Serialize)]
+pub enum SupportResult {
+    /// The standard is not supported.
+    NoSupport,
```

Comment:
I prefer `NoSupport`, because I feel it fits better, unless you have some argument for changing it, I'll leave it as is.

- I provide no support for this.
- I provide support for this.
- These contracts provide my support for this.